### ISSUE TITLE
Change data/dataKind fields to be a map (kind -> data) instead

### ISSUE THREAD

**Issue by agluszak** [build-server-protocol/build-server-protocol#458]
This will allow build targets and other structures which are allowed to carry additional data to include more than one kind of data. For example, [Scala targets](https://build-server-protocol.github.io/docs/extensions/scala#scala-build-target) need both Scala-specific data and more general JVM data. Currently that's solved by nesting the second kind of data as a field in the first kind of data. This won't scale well when we start supporting more extensions to the base protocol.

**Comment by adpi2** [comment]
:+1: This would solve a problem that I have currently.

I would like to add a Twirl extension in BSP but currently this is only possible either by nesting the Scala extension into the Twirl one, or the Twirl one into the Scala one, which I would try to avoid both. A map of kind to data would be an elegant solution.

**Comment by jastice** [comment]
Agree. Essentially we could define `data` simply as a a json object with keys/values in the protocol.
Question is only how to avoid naming conflicts. Perhaps recommending namespaces is sufficient though.

**Comment by adpi2** [comment]
> Question is only how to avoid naming conflicts. Perhaps recommending namespaces is sufficient though.

Not sure what you mean by `namespaces`. Something like this should work though:
```
"data":{
  "jvm": {
    "javaHome": "???",
    "javaVersion": "???"
  },
  "scala":  {
    "scalaVersion": "???",
    "jars": ["", ""];
  },
 "twirl": {
    "twirlVersion": "???",
    "twirlCompilerJar": "???"
  }
}
```

**Comment by jastice** [comment]
I meant that if the number of data kinds becomes large enough, people might name their fields the same inadvertently. In which case we'd want something like the java package names convention. This doesn't seem necessary at the moment though.

Separately, we could support this in BSP 2.x by introducing a `datakind: "map"` that specifies that `data` can have multiple fields, and their content corresponds to the existing shapes for `data`. What do you think?

**Comment by adpi2** [comment]
I agree that we should support this in BSP 2.x, by introducing a `"datakind": "map"` (or `"datakind": "dictionary"`). However I think we should not flatten the content of existing shapes in the same `data` object because, as you raised, there can be conflicts between keys, and also because the client would have to guess what data kind is provided. I think a better approach is to add one layer of nesting, like this:
```ts
{
  "datakind": "dictionary",
  "data": {
     "jvm": {
        // the fields of the JVM build target
      },
      "scala": {
        // the fields of the Scala build target
        // jvmBuildTarget should be left empty (it is optional)
      },
     "java": {
        // the fileds of the Java build target
     }
  }
}
```

Here the keys of `data` are some defined `datakind`, and their values are the corresponding build targets.

**Comment by lefou** [comment]
Having a data map _and_ embedded data brings potential for duplication, consistency issues and conflicts. E.g. a scala module may contain some java and jvm settings and it may as well be a twirl module which also embeds some scala and jvm settings. Some good guidelines are needed then, when to refer to another data and when to embed. Better, we avoid such construct from the beginning and only allow a flat data map. Instead of embedding data, we just refer to the keys in the map?

**Comment by jastice** [comment]
@lefou just to clarify, you mean something like:

```
"data":{
    "javaHome": "???",
    "javaVersion": "???"
    "scalaVersion": "???",
    "jars": ["", ""];
    "twirlVersion": "???",
    "twirlCompilerJar": "???"
}
```

instead of

```
"data":{
  "jvm": {
    "javaHome": "???",
    "javaVersion": "???"
  },
  "scala":  {
    "scalaVersion": "???",
    "jars": ["", ""];
  },
 "twirl": {
    "twirlVersion": "???",
    "twirlCompilerJar": "???"
  }
}
```

?

**Comment by adpi2** [comment]
```
"data":{
    "javaHome": "???",
    "javaVersion": "???"
    "scalaVersion": "???",
    "jars": ["", ""];
    "twirlVersion": "???",
    "twirlCompilerJar": "???"
}
```

That could work, but it would be nice then if the `datakind` lists all the data kinds instead of just `map`.

Like this: `"datakind": "scala,jvm,java,twirl"`.
Or this: `"datakind": ["scala", "jvm", "java", "twirl"]`.

**Comment by lefou** [comment]
@adpi2 wrote:
> ```ts
> {
>   "datakind": "dictionary",
>   "data": {
>      "jvm": {
>         // the fields of the JVM build target
>       },
>       "scala": {
>         // the fields of the Scala build target
>         // jvmBuildTarget should be left empty (it is optional)
>       },
>      "java": {
>         // the fileds of the Java build target
>      }
>   }
> }
> ```
> 
> Here the keys of `data` are some defined `datakind`, and their values are the corresponding build targets.

This is also my preferred mental model, but I think the fact that we keep the nested/embedded "jvmBuildTarget" data in the scala data is problematic. Take the following example:


```ts
{
  "datakind": "dictionary",
  "data": {
     "jvm": {
        // the fields of the JVM build target
      },
      "scala": {
        // the fields of the Scala build target
        "jvmBuildTarget": {
          // some fields of the JVM build target
       }
      },
     "java": {
        // the fileds of the Java build target
     }
  }
}
```

Which is the correct JVM data? The top-level "jvm" or the nested "jvmBuildTarget" (in the scala data) one? What if they differ? Should they be merged or override each other? Should the nested one used only in some (scala) cases? What if a client does not support one of the data types in the map, can it assume that the top-level "jvm" data is complete?

IMHO, we should change the current spec and remove nested data. If not, we should be very specific, how this all is supposed to work. We need guidelines for server and client implementers then.

**Comment by abrams27** [comment]
> That could work, but it would be nice then if the datakind lists all the data kinds instead of just map.

> Like this: "datakind": "scala,jvm,java,twirl".
Or this: "datakind": ["scala", "jvm", "java", "twirl"].

but if it's a map we dont need the `dataKind` at all - the map keys should do the trick

**Comment by adpi2** [comment]
> but if it's a map we dont need the dataKind at all - the map keys should do the trick

Do you mean that we should force it to be a map, always?

**Comment by abrams27** [comment]
```
"data":{
    "javaHome": "???",
    "javaVersion": "???"
    "scalaVersion": "???",
    "jars": ["", ""];
    "twirlVersion": "???",
    "twirlCompilerJar": "???"
}
```
in this case we can read all the available fields from the map
